source/common/common/callback_impl.h

```diff
@@ -81,5 +106,72 @@ template <typename... CallbackArgs> class CallbackManager {
   const std::shared_ptr<bool> still_alive_{std::make_shared<bool>(true)};
 };
 
+/**
+ * @brief Utility class for managing callbacks across multiple threads.
+ *
+ * @see CallbackManager for a non-thread-safe version
+ */
+class ThreadSafeCallbackManager : public std::enable_shared_from_this<ThreadSafeCallbackManager> {
+  struct CallbackHolder;
+  using CallbackListEntry = std::tuple<CallbackHolder*, Event::Dispatcher&, std::weak_ptr<bool>>;
+
+public:
+  using Callback = std::function<void()>;
+
+  /**
+   * @brief Create a ThreadSafeCallbackManager
+   *
+   * @note The ThreadSafeCallbackManager must always be represented as a std::shared_ptr in
+   *       order to satisfy internal conditions to how callbacks are managed.
+   */
+  static std::shared_ptr<ThreadSafeCallbackManager> create() {
+    return std::shared_ptr<ThreadSafeCallbackManager>(new ThreadSafeCallbackManager());
+  }
+
+  /**
+   * @brief Add a callback.
+   * @param dispatcher Dispatcher from the same thread as the registered callback. This will be used
+   *                   to schedule the execution of the callback.
+   * @param callback callback to add
+   * @return Handle that can be used to remove the callback.
+   */
+  ABSL_MUST_USE_RESULT CallbackHandlePtr add(Event::Dispatcher& dispatcher, Callback callback);
+
+  /**
+   * @brief Run all callbacks
+   */
+  void runCallbacks();
+
+  size_t size() const noexcept;
+
+private:
+  struct CallbackHolder : public CallbackHandle {
+    CallbackHolder(std::weak_ptr<ThreadSafeCallbackManager> parent, Callback cb,
```

Comment:
Using shared_ptr instead of a weak_ptr would result in simpler code.

Reply:
Agreed. Done.